src/sources/envoy_als/grpc.rs

```diff
@@ -0,0 +1,136 @@
+use value::Value;
+use vector_common::internal_event::{
+    ByteSize, BytesReceived, CountByteSize, EventsReceived, InternalEventHandle as _, Registered,
+};
+use vector_core::{event::LogEvent, EstimatedJsonEncodedSizeOf};
+
+#[derive(Clone)]
+pub(super) struct Service {
+    pub events_received: Registered<EventsReceived>,
+    pub bytes_received: Registered<BytesReceived>,
+    pub pipeline: SourceSender,
+    pub shutdown: ShutdownSignal,
+}
+
+enum StreamError {
+    Grpc(Status),
+    StreamClosed,
+}
+
+#[tonic::async_trait]
+impl AccessLogService for Service {
+    async fn stream_access_logs(
+        &self,
+        request: Request<Streaming<StreamAccessLogsMessage>>,
+    ) -> Result<Response<StreamAccessLogsResponse>, Status> {
+        let mut in_stream = request.into_inner();
+        let mut shutdown = self.shutdown.clone();
+        let mut stream_identifier = None;
+
+        loop {
+            tokio::select! {
+                _ = &mut shutdown => break,
+                stream_msg = in_stream.message() => {
+                    if let Err(err) = self.handle_message(stream_msg, &mut stream_identifier).await {
+                        match err {
+                            StreamError::StreamClosed => break,
+                            StreamError::Grpc(status) => {
+                                emit!(GrpcError{error: status.clone()});
+                                return Err(status);
+                            }
+                        }
+                    }
+                },
+            }
+        }
+
+        Ok(Response::new(StreamAccessLogsResponse {}))
+    }
+}
+
+impl Service {
+    async fn handle_message(
+        &self,
+        stream_msg: Result<Option<StreamAccessLogsMessage>, Status>,
+        stream_identifier: &mut Option<Identifier>,
+    ) -> Result<(), StreamError> {
+        match stream_msg {
+            Ok(omsg) => match omsg {
+                Some(msg) => {
+                    self.process_logs(msg, stream_identifier).await?;
+                }
+                None => {
+                    // Envoy gracefully closed the stream
+                    return Err(StreamError::StreamClosed);
+                }
+            },
+            Err(err) => {
+                return Err(StreamError::Grpc(err));
+            }
+        }
+        Ok(())
+    }
+
+    async fn process_logs(
+        &self,
+        msg: StreamAccessLogsMessage,
+        stream_identifier: &mut Option<Identifier>,
+    ) -> Result<(), StreamError> {
+        let mut events = vec![];
+
+        // identifier is only sent on the first message of a stream,
+        // but we want to have it alongside all log events so they can be
+        // identified by the Envoy instance that sent them. To do this,
+        // we store the identifier sent on the stream and tack it on to
+        // future events.
+        if let Some(identifier) = msg.identifier {
+            *stream_identifier = Some(identifier);
+        }
```

Comment:
Is the stream identifier possibly containing sensitive data?
I think it might be helpful to have an info log stating that the identifier was received (since it only would be happening once). If it doesn't contain sensitive data we could also print the identifier in the console log message.

Reply:
Sure. I'm thinking now maybe we should convert this to the Vector `Value` and store that in memory rather than the proto message. Then log the that instead. The reason is, I cut one of the fields on the identifier (extensions) due to it being large and not super useful (contains a list of every extensions available in the version of Envoy that connected, which is >100 items)